src/modm/platform/spi/sam_x7x/spi_hal.hpp.in

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2023, Christopher Durand
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_SAMX7X_SPI_HAL{{ id }}_HPP
+#define MODM_SAMX7X_SPI_HAL{{ id }}_HPP
+
+#include <cstddef>
+#include <cstdint>
+#include "spi_base.hpp"
+
+namespace modm::platform
+{
+
+/**
+ * Serial peripheral interface (SPI{{ id }})
+ *
+ * Low-level SPI peripheral register access API
+ *
+ * @author		Christopher Durand
+ * @ingroup		modm_platform_spi modm_platform_spi_{{id}}
+ */
+class SpiHal{{ id }} : public SpiBase
+{
+public:
+	/// Enable the clock, reset and enable the peripheral
+	static void
+	enable();
+
+	/// Disable the peripheral clock
+	static void
+	disable();
+
+	/**
+	 * Enable and initialize SPI in master mode
+	 *
+	 * \param prescaler clock divider (1..255), value 0 is not allowed
+	 * \todo Support for slave mode and hardware-managed chip select
+	 */
+	static void
+	initialize(uint8_t prescaler,
+				DataMode dataMode = DataMode::Mode0,
+				DataSize dataSize = DataSize::Bit8);
+
+	static void
+	setDataMode(DataMode dataMode);
+
+	static void
+	setDataSize(DataSize dataSize);
+
+	static void
+	setMasterSelection(MasterSelection masterSelection);
+
+	static void
+	setLoopbackMode(LocalLoopback loopback);
+
+	/**
+	 * Write up to 8 bits to the data register
+	 *
+	 * \pre Transmit data register must be empty
+	 */
+	static void
+	write(std::byte data);
```

Comment:
`std::byte` is the recommended type to represent the concept of binary data since C++17. It is a distinct type that is neither a character type nor an arithmetic type but you can use bit-wise operations on it and explicitly convert it to integer. It shares the strict aliasing exceptions of `char`, so you can legally `reinterpret_cast` any valid object pointer to `std::byte*` and dereference it.
It also does not have the integer promotions from C. The bit-or of two `std::byte` is still a byte, not like `char`s which give you `int`.
If you prefer to use `uint8_t` to be consistent with older code, I can change it to that.

Reply:
Oh wow, they finally fixed it! No, keep it like this, we should probably update all interfaces in the future though.